emission/analysis/classification/inference/labels/pipeline.py

```diff
@@ -0,0 +1,155 @@
+def placeholder_predictor_2(trip):
+    global trip_n  # ugly use of globals for testing only
+    trip_n %= 6
+    trip_n += 1
+    return [
+        [
+
+        ],
+        [
+            {"labels": {"mode_confirm": "bike", "purpose_confirm": "work"}, "p": 0.8},
+            {"labels": {"mode_confirm": "walk", "purpose_confirm": "shopping"}, "p": 0.2}
+        ],
+        [
+            {"labels": {"mode_confirm": "drove_alone"}, "p": 0.8},
+        ],
+        [
+            {"labels": {"mode_confirm": "bike", "purpose_confirm": "work"}, "p": 0.8},
+            {"labels": {"mode_confirm": "walk", "purpose_confirm": "shopping"}, "p": 0.2}
+        ],
+        [
+            {"labels": {"mode_confirm": "walk", "purpose_confirm": "shopping"}, "p": 0.45},
+            {"labels": {"mode_confirm": "walk", "purpose_confirm": "entertainment"}, "p": 0.35},
+            {"labels": {"mode_confirm": "drove_alone", "purpose_confirm": "work"}, "p": 0.15},
+            {"labels": {"mode_confirm": "shared_ride", "purpose_confirm": "work"}, "p": 0.05}
+        ],
+        [
+            {"labels": {"mode_confirm": "walk", "purpose_confirm": "shopping"}, "p": 0.45},
+            {"labels": {"mode_confirm": "walk", "purpose_confirm": "entertainment"}, "p": 0.35},
+            {"labels": {"mode_confirm": "drove_alone", "purpose_confirm": "work"}, "p": 0.15},
+            {"labels": {"mode_confirm": "shared_ride", "purpose_confirm": "work"}, "p": 0.05}
+        ]
+    ][6-trip_n]
+
+# For each algorithm in ecwl.AlgorithmTypes that runs on a trip (e.g., not the ensemble, which
+# runs on the results of other algorithms), primary_algorithms specifies a corresponding
+# function to run. This makes it easy to plug in additional algorithms later.
+primary_algorithms = {
+    # This can be edited to select a different placeholder predictor
+    ecwl.AlgorithmTypes.PLACEHOLDER: placeholder_predictor_2
+}
+
+# Code structure based on emission.analysis.classification.inference.mode.pipeline
+# and emission.analysis.classification.inference.mode.rule_engine
+class LabelInferencePipeline:
+    def __init__(self):
+        self._last_trip_done = None
+    
+    @property
+    def last_trip_done(self):
+        return self._last_trip_done
+
+    # For a given user and time range, runs all the primary algorithms and ensemble, saves results
+    # to the database, and records progress
+    def run_prediction_pipeline(self, user_id, time_range):
+        self.ts = esta.TimeSeries.get_time_series(user_id)
+        self.toPredictTrips = esda.get_entries(
+            esda.CONFIRMED_TRIP_KEY, user_id, time_query=time_range)
+        for trip in self.toPredictTrips:
+            results = self.compute_and_save_algorithms(trip)
+            ensemble = self.compute_and_save_ensemble(trip, results)
+
+            # Add final prediction to the confirmed trip entry in the database
+            trip["data"]["inferred_labels"] = ensemble["prediction"]
+            self.ts.update(trip)
```

Comment:
I saw your question about this. In general, we have had the input to each pipeline stage be read-only, so each stage generates new output. That is why we avoid `ts.update`. This made sense when the outputs could be significantly different from the inputs (e.g. the cleaned trip is the segmented trip + some extrapolation at the start). 

But it gets a bit more complicated when we are just changing a few fields in the object; creating a new object (e.g. `labeled_trip`) is going to require rewriting the client as well to read the new kind of object.

If we did want to go down that route, we would create a new `labeled_trip` or `inferred_labeled_trip` datastructure, which would include inferred labels but not user inputs and be the output of this stage.

we would then create `confirmed_trip` objects from the `inferred_labeled_trip` instead of `cleaned_trip`. Now that I've written it out, maybe that is cleaner after all than modifying the `confirmed_trip` directly.

It retains the high level assumptions of the pipeline and does not require client changes.

Thoughts?

Reply:
I agree that we could do it without `ts.update` reasonably cleanly and without requiring client changes by adding a new trip data structure. It does seem a bit overkill to copy the entire trip just to add a single field, but I'm not too well-versed in "classic ML algorithms"; if this is an established standard, maybe we should follow it.

Reply:
Maybe more "classic ML systems" - e.g. Spark et al. Making the data structures read-only allows you to maintain genealogy and roll forward computations efficiently.  How big is each trip structure? Unlike sections, they don't store the actual trajectory-related information, so I would assume that they are fairly lightweight.

Reply:
You're right, the `repr()` of each confirmed trip is on the order of 1500 characters.